### ISSUE TITLE
[LOW] [Splits] [$500] IOU - When paying a deleted IOU disappears and RBR does not appear when going online

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#38338]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.52-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4424536
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Open NewDot app or https://staging.new.expensify.com/
2. Log in to account A and B in incognito mode
3. [User B] Request money from User A
4. [User A] Go offline
5. [User A] Open the IOU report and tap Pay elsewhere
6. [User B] Delete the IOU
7. [User A] Go online

## Expected Result:
The deleted IOU should not disappear when you pay and should appear RBR when you go online

## Actual Result:
When paying a deleted IOU disappears and RBR does not appear when going online

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/93399543/7933a1bd-6eb5-48ef-9ff0-25f3404b2ac3

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01606c57f19fab1b0c</li>
        <li>Upwork Job ID: 1768749861959938048</li>
        <li>Last Price Increase: 2024-08-08</li>
    </ul>
</details>

**Comment by kbecciv** [comment]
@mallenexpensify I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors.

**Comment by kbecciv** [comment]
We think that this bug might be related to #vip-bills
CC @quinthar

**Comment by mallenexpensify** [comment]
Was able to reproduce, pretty sure we're filing this in #vip-split. Going to add there.  
@akinwale I marked this External, do you agree?

**Comment by akinwale** [comment]
> @akinwale I marked this External, do you agree?

Yes. It makes sense to let the user know that the attempted payment actually failed.

**Comment by ijmalik** [comment]
**Please re-state the problem that we are trying to solve in this issue.**

1. Open NewDot app or https://staging.new.expensify.com/
2. Log in to account A and B in incognito mode
3. [User B] Request money from User A
4. [User A] Go offline
5. [User A] Open the IOU report and tap Pay elsewhere
6. [User B] Delete the IOU
7. [User A] Go online
8. Paid  IOU disappears and RBR does not appear when going online

**What is the root cause of that problem?**
In Step-6 [User B] has deleted the IOU i.e. IOU is no more on server.
When [User A] Go online, API call send PayMoneyRequest and because related paid IOU is no more on server so server replied with following response with error message

```
{
    "code": 666,
    "jsonCode": 666,
    "type": "Expensify\\Libs\\Error\\ExpError",
    "UUID": "85427d9f-f33a-47af-ae93-2249272d9fa6",
    "message": "An unexpected error occurred. Please try again.",
    "title": "",
    "data": {
        "onyxData": [
            {
                "onyxMethod": "merge",
                "key": "reportActions_6234418514202091",
                "value": {
                    "3868470631178296430": {
                        "errors": {
                            "1710911738535230": "An unexpected error occurred. Please try again."
                        }
                    }
                }
            }
        ]
    },
    "htmlMessage": "",
    "onyxData": [
        {
            "onyxMethod": "merge",
            "key": "reportActions_6234418514202091",
            "value": {
                "3868470631178296430": {
                    "errors": {
                        "1710911738535230": "An unexpected error occurred. Please try again."
                    }
                }
            }
        }
    ],
    "requestID": "867339bcb8ccd1e8-MCT"
}
```
This error message should be shown to user but Server IOU deleted data updated at user side which delete the IOU for [User A] also. That's why IOU disappears and RBR not shown

**What changes do you think we should make in order to solve the problem?**
In https://github.com/Expensify/App/blob/main/src/libs/Middleware/SaveResponseInOnyx.ts we have all server  responses and our requests here. We shall do following steps
1- Trace error response of PayMoneyRequest from server
2. Manage / change response accordingly to show required info at [User A] side
3- We also trace server response which deletes IOU and prevent it to delete IOU from User side.



https://github.com/Expensify/App/assets/2220885/130fd814-ccde-4eb1-a3f8-eebf0ab09124

**Comment by mallenexpensify** [comment]
@akinwale can you please review the proposal above? Thx

**Comment by mallenexpensify** [comment]
Removed @akinwale , was ten days without a review of the proposal.  @abdulrahuman5196 , can you please review the above? https://github.com/Expensify/App/issues/38338#issuecomment-2014433112

**Comment by abdulrahuman5196** [comment]
Hi, will review this today.

**Comment by abdulrahuman5196** [comment]
On @ijmalik 's proposal here - https://github.com/Expensify/App/issues/38338#issuecomment-2014433112. 
I don't agree on the solution suggested to make changes in the central SaveResponseInOnyx.ts component to fix this particular case. 
I think we don't have any proposals to approve yet.

**Comment by abdulrahuman5196** [comment]
We don't have proposals pending for review .

**Comment by abdulrahuman5196** [comment]
We don't have proposals to review.

**Comment by mallenexpensify** [comment]
Throwing `retest-weekly` on here since it's been open a while.

**Comment by abdulrahuman5196** [comment]
Waiting on retest

**Comment by mallenexpensify** [comment]
Waiting on retest (leaving a daily cuz it should be retested soon)

**Comment by mallenexpensify** [comment]
Checking on process and timeline for 'retest-weekly' label in #qa https://expensify.slack.com/archives/C9YU7BX5M/p1714532562694299

**Comment by mvtglobally** [comment]
Issue is reproducible during KI retests. 

https://github.com/Expensify/App/assets/43995119/ade31de8-8397-4e5a-aa1c-328e626d5873

**Comment by mallenexpensify** [comment]
We have site issues and deploy freeze, I'll revisit next week.

**Comment by mallenexpensify** [comment]
Waiting on proposals, since it's [LOW], I'm not eager to raise the price about $500 right now.

**Comment by mallenexpensify** [comment]
Still waiting

**Comment by wildan-m** [comment]
I believe the current behavior is as expected, with the deletion process received by the API server first and the payment being sent later when the user online. Or I missed some steps?

**Comment by mallenexpensify** [comment]
@wildan-m , I'm unsure.  @abdulrahuman5196 , do you know?

**Comment by abdulrahuman5196** [comment]
>I believe the current behavior is as expected, with the deletion process received by the API server first and the payment approval action being sent later when the user online. Or I missed some steps?

> @wildan-m , I'm unsure. Thanks for the comment. @abdulrahuman5196 , do you know?

@mallenexpensify 
Can we confirm this experience with someone else as well if you are unsure?
Because the above mentioned is the steps from the dev perspective. But in terms of customer perspective, customer marked a request as paid, but it suddenly disappeared without showing error or anything. So IMO this is a bug, but would be good to confirm as well.

**Comment by mallenexpensify** [comment]
Here's the TestRail from where the bug was found. 
<img width="948" alt="image" src="https://github.com/Expensify/App/assets/22508304/c30944c5-40d6-4280-a5f2-fc6ef0f67a8d">

> the deletion process received by the API server first and the payment approval action being sent later when the user online

Even if this is the case, it appears this would still be a bug if it's not functioning as intended, based on the TestRail case above.  I'm unsure how best to fix though

**Comment by ijmalik** [comment]
## Proposal
[Updated]https://github.com/Expensify/App/issues/38338#issuecomment-2014433112

**Comment by mallenexpensify** [comment]
@abdulrahuman5196 , 👀 on @ijmalik 's proposal above plz

**Comment by abdulrahuman5196** [comment]
> On @ijmalik 's proposal here - [#38338 (comment)](https://github.com/Expensify/App/issues/38338#issuecomment-2014433112). I don't agree on the solution suggested to make changes in the central SaveResponseInOnyx.ts component to fix this particular case. I think we don't have any proposals to approve yet.

I am still on this. I don't think its a good solution to update the central middlewares and API infra to reflect this exception.

Currently we don't have any proposals to be approved yet.

**Comment by abdulrahuman5196** [comment]
@mallenexpensify Can we involve an internal engineer here to provide information from backend as well? Or some other direction as well.

Basically, the reportAction(IOU) is deleted from the ReconnectApp API call response. So we don't have a report action where we can show an error.

**Comment by mallenexpensify** [comment]
@abdulrahuman5196 before assigning an engineer, can you post in #expensify-open-source to see if you can get a reply from someone from the team?  Thx

**Comment by sonialiap** [comment]
@mallenexpensify @abdulrahuman5196 I came across a similar sounding issue, do you think they're connected https://github.com/Expensify/App/issues/40705

**Comment by mallenexpensify** [comment]
Good 👀 @sonialiap , the steps are near identical. @abdulrahuman5196 can you review that issue and comment here or there with any thoughts?  Normally we'd keep the older issue open but it looks like both these have quite a few comments (and no suggested proposals).

**Comment by abdulrahuman5196** [comment]
@mallenexpensify Both are almost similar issues. And I can see comments that other issue is not reproducible. So we can focus on this issue. 

But anyways I agree on this https://github.com/Expensify/App/issues/40705#issuecomment-2176587187, we need internal guidance on `if we should fix this`, and if so in what direction.

**Comment by mallenexpensify** [comment]
Checking in #vip-split to see what folks think
https://expensify.slack.com/archives/C05RECHFBEW/p1720481477873549

**Comment by mallenexpensify** [comment]
From [chat a couple weeks ago](https://expensify.slack.com/archives/C05RECHFBEW/p1720513768742349?thread_ts=1720481477.873549&cid=C05RECHFBEW)
> I think we should fix this bug but not now. You can make it a monthly and add it under the paused items under this proje

I removed `Help Wanted`

**Comment by mallenexpensify** [comment]
Gonna be a min til we get to this

**Comment by mallenexpensify** [comment]
IOUs are a priority right now

**Comment by mallenexpensify** [comment]
Closing as we're closing #vip-split
